Report type of expressions returning polyglot values

hubertp opened the pull request:
### Pull Request Description

Expressions returning polyglot values were not reporting the type of the result because we have to do additional magic that infers the correct Enso type. Since this is exactly what `TypeOfNode` does, I re-used the logic.

Straightforward solution failed in tests because of assertions:
```
[enso] WARNING: Execution of function main failed (Invalid library usage. Cached library must be adopted by a RootNode before it is executed.).
java.lang.AssertionError: Invalid library usage. Cached library must be adopted by a RootNode before it is executed.
```

That is why this PR replaces `ExecutionEventListener` with `ExecutionEventNodeFactory`.

### Important Notes

Usage of `TypeOfNode` for programs that **do not** import stdlib means that we report types that do not involve stdlib e.g.
`Standard.Builtins.Main.Integer` instead of `Standard.Base.Data.Numbers.Integer`. While surprising, this is correct and I would say desirable. While reviewing the code, notice the difference in expectations in our runtime tests.

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Report type of expressions returning polyglot values

Expressions returning polyglot values were not reporting the type of the
result because we have to do additional magic that infers the correct
Enso type. Since this is exactly what `TypeOfNode` does, I re-used the
logic.

Except this doesn't work in tests because they have assertions on and we
get
```
[enso] WARNING: Execution of function main failed (Invalid library usage. Cached library must be adopted by a RootNode before it is executed.).
java.lang.AssertionError: Invalid library usage. Cached library must be adopted by a RootNode before it is executed.
```

JaroslavTulach commented on engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java:
```diff
@@ -64,6 +66,7 @@ private static class IdExecutionEventListener implements ExecutionEventListener
     private final UUID nextExecutionItem;
     private final Map<UUID, FunctionCallInfo> calls = new HashMap<>();
     private final Timer timer;
+    private final TypeOfNode typeOfNode;
```
Using node in an instrument is completely fine. But instead of `ExecutionEventListener` we should rewrite the code to [ExecutionEventNodeFactory](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/instrumentation/ExecutionEventNodeFactory.html) - then the `TypeOfNode` will get _"mounted"_ (and adopted) in the tree of other nodes.

JaroslavTulach commented on engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java:
```diff
@@ -187,6 +191,13 @@ private void onExpressionReturn(Object result, Node node, EventContext context)
         String resultType = Types.getName(result);
+        if (resultType == null && result != null) {
+            // Slow path
+            Object typeResult = typeOfNode.execute(result);
```
The whole method is behind `@TruffleBoundary` right now. I suggest to change it and call `typeOfNode.execute` in PE mode. It should be quite fast.

hubertp committed:
s/ExecutionEventListener/ExecutionEventNodeFactory

Replacing ExecutionEventListener with ExecutionEventNodeFactory allows
us to enhance EventNode with an additional child node, TypeOfNode. The
latter gets adopted to the RootNode, meaning it won't blow up assertions
anymore.

Additionally reduced the scope of @TruffleBoundary so that the child can
possibly get PE, as suggested by Jaroslav.

hubertp committed:
Replace Types.getName with TypeOfNode

Rather than introducing a fallback case with `TypeOfNode` this change
replaces the `Types.getName` completely.
This has a number of consequences:
- if one does not import stdlib, all those types have a builtin type
- the type of atom constructor is no longer the fqn name of the atom but
  a Function
- unresolved_symbol has to be treated specially

All those changes, while surprising, are actually correct. I also
believe they are desirable but I'm open for discussion.

hubertp committed:
Update CHANGELOG

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java:
```diff
@@ -50,6 +54,21 @@ Object doString(String value) {
+
+  @Specialization
+  Object doWarning(WithWarnings value) {
+    return execute(value.getValue());
```
@JaroslavTulach The type of the Warning is the type of the underlying value - that is the expected semantics, right? Nothing broke because of that so I'm assuming this is OK.

hubertp committed:
Remove remaining usage of Tests.getName

hubertp commented on engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/BuiltinTypesTest.scala:
```diff
@@ -319,7 +319,7 @@ class BuiltinTypesTest
     ) should contain theSameElementsAs Seq(
       Api.Response(requestId, Api.PushContextResponse(contextId)),
-      TestMessages.update(contextId, idMain, "Enso_Test.Test.Main.Foo.Bar"),
+      TestMessages.update(contextId, idMain, ConstantsGen.FUNCTION_BUILTIN),
```
This might seem like a weird change at first sight but consider that `Bar` is a constructor taking an argument. So the type of it is a Function. `Types.getName` had a special case for `AtomConstructor` and I think now we are more consistent.

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java:
```diff
@@ -77,8 +77,10 @@ public void checkAllConstantGenValuesArePresent() throws Exception {
     for (var f : ConstantsGen.class.getFields()) {
-      var s = (String) f.get(null);
-      expecting.add(s);
+      if (!f.getName().endsWith("_BUILTIN")) {
```
What is this?

hubertp replied:
This is so that we don't have to write (by hand) FQN of builtin types. We generate `ConstantsGen` for types from stdlib but if one does not import stdlib builtin types are essentially `Standard.Builtin.Main.XYZ`. So every type which is in stdlib, say `Standard.Base.Foo.Bar.XYZ` has an equivalent in `Standard.Builtin.Main.XYZ`. Rather than trying to keep those up-to-date, processor adds those to `ConstantsGen`.

You might ask why do we need this at all?
Well, in tests (but also in a real application), if you don't import stdlib you operate only on builtins and the reported type shouldn't really be some type from stdlib, should it? That is why the type reported by `TypeOfNode` is an actual type, not some hardcoded string (like `Types.getName`). Hence the changes in tests (the changed ones don't import stdlib). 

Also, the test assumes that stdlib is imported. It does not operate on builtins.

JaroslavTulach reviewed (approved):
Usage of `ExecutionEventNodeFactory` seems correct. Using nodes from instruments is a good thing - and it is also fast. Sharing such _"utility nodes"_ like `TypeOfNode` at various places is the way to go.

hubertp committed:
Merge branch 'develop' into wip/hubert/infer-typeof-polyglot-values-184336928

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/infer-typeof-polyglot-values-184336928

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/infer-typeof-polyglot-values-184336928

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/infer-typeof-polyglot-values-184336928

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/infer-typeof-polyglot-values-184336928
